src/h3/mod.rs

```diff
@@ -642,19 +650,47 @@ impl Connection {
     /// a newly allocated stream.
     ///
     /// On success the newly allocated stream ID is returned.
+    ///
+    /// The [`StreamBlocked`] error is returned when the underlying QUIC stream
+    /// doesn't have enough capacity for the operation to complete. When this
+    /// happens the application should retry the operation once the stream is
+    /// reported as writable again.
+    ///
+    /// [`StreamBlocked`]: enum.Error.html#variant.StreamBlocked
     pub fn send_request(
         &mut self, conn: &mut super::Connection, headers: &[Header], fin: bool,
     ) -> Result<u64> {
-        let stream_id = self.get_available_request_stream()?;
+        let stream_id = self.next_request_stream_id;
+
         self.streams
             .insert(stream_id, stream::Stream::new(stream_id, true));
 
+        // The underlying QUIC stream does not exist yet, so calls to e.g.
+        // stream_capacity() will fail. By writing a 0-length buffer, we force
+        // the creation of the QUIC stream state, without actually writing
+        // anything.
+        conn.stream_send(stream_id, b"", false)?;
+
         self.send_headers(conn, stream_id, headers, fin)?;
 
+        // To avoid skipping stream IDs, we only calculate the next available
+        // stream ID when a request has been successfully buffered.
+        self.next_request_stream_id = self
+            .next_request_stream_id
+            .checked_add(4)
+            .ok_or(Error::IdError)?;
```

Comment:
oh this reminded me of the latent bug where effective maximum stream ID is not u64::MAX but rather 2^62-1. But the chances of either being hit are very small :)

Reply:
I guess the transport will end up returning an error in that case, so the client will keep failing to send a request, which is probably as good as it can get.